### PR TITLE
[FIX] website: fix website page properties

### PR DESCRIPTION
**Pull request by xO-Tx**
The goal of this commit is to add some fixes on page properties dialog after the Owl REF [1].

- XML: use default form view style (remove `</group>`).
- Move the '/' back into the non editable part of the URL.
- Prevent python code from creating a different URL (and optionally setting `website.rewrite`
record for it) when the new URL is the same as the initial one after slugify.
- '`useAutofocus()`' on the first page properties field.

[1]: https://github.com/odoo/odoo/commit/61a9d7bd2abc6081b321a51d329f9e85209215e5

task-2687506

### CODE REVIEW

**Comment by xO-Tx** [comment]
@qsm-odoo a supplementary fix for page properties :+1:
Just a remark about UI: since the [fields CSS update](https://github.com/odoo/odoo/commit/77824ad44b6945a9811120380747f87ef6362ae2#diff-c6cbc1080c8fd8cefd4a2334f5be5a326ec4826c9ca979358b4314ee1402a49eR7), there is no border for fields on form views (including page properties...)

16 0-field-border-onhover-mou

Not sure if this is intended to be this way :thinking:

File: addons/website/static/src/components/fields/fields.js [hunk 1]
```diff
@@ -16,26 +16,49 @@ const {Component, useState, onWillStart} = owl;
...
+                    const extension = section.slice(i);
+                    if (i > 0 && this.pageMimetypes[extension]) {
+                        slugified = _.str.slugify(section.slice(0, i)) + extension;
+                    }
+                }
+                return slugified ? `${path}/${slugified}` : path;
+            }, '');
+        }
+        return url.replace(/^\//g, '');
+    }
```
Comment (qsm-odoo) [inline addons/website/static/src/components/fields/fields.js:60]: Could you use some comments and function documentations. Was the "slugification" done in JS like that in 15.0? Trying to remember how this got lost :) cc @rdeodoo Maybe that was done in python but finally doing it in JS is ok because the python lead to more code something like that?

**Review by qsm-odoo — COMMENTED** [review]
> Just a remark about UI: since the [fields CSS update](https://github.com/odoo/odoo/commit/77824ad44b6945a9811120380747f87ef6362ae2#diff-c6cbc1080c8fd8cefd4a2334f5be5a326ec4826c9ca979358b4314ee1402a49eR7), there is no border for fields on form views (including page properties...)

16 0-field-border-onhover-mou

> Not sure if this is intended to be this way :thinking:

I guess it's indented. Although maybe worth a ping to @bouvyd : I have still yet to try the new "edit at all times" feature in details (seemed nice for the 5min I took to test it out) but in this case where it is in a dialog you enter with no other purpose than editing... it feels weird. Or do you think it's something to get used to there too? (You can try it out on this branch with Website -> Site -> Properties). Notice that "website expertise people" sent us a document where they also report this as a bug (I'll send you the link on Discord) :grimacing:

File: addons/website/static/src/components/fields/fields.js (hunk 1 above)
Comment (rdeodoo) [reply addons/website/static/src/components/fields/fields.js:60]: The slugify is always done in python and always come with the unique path check, slugifying in js will always be incomplete and there is no way to compute the final url, as it needs to reach all other pages to ensure it is not already taken (and if so act accordingly, something like adding a suffix, I don't have code in front of me right now)

> Add more accuracy when comparing user input to the current page url
to prevent some situations where the redirect is useless:
Example: old page url value = '/home-page' and user value = 'home page'
=> this will lead to the same url = '/home-page' (after slugify on
backend) so the redirect here doesn't make any sense and the redirect
option should not be visible.

I am not sure why would someone rename the current page url "/contact-us" to "contact us".

Why would we want code for handling that? That's overkill optimization, this should be like 1 case out of 10.000 url modification, on top of being weird to do. In that case we shouldn't even care about the redirect (which is not even correct in js anyway as I explained).

Just a quick reply without being able to check the code or blame anything, based on the commit message. So it might have been like that before but I doubt it.

Comment (qsm-odoo) [reply addons/website/static/src/components/fields/fields.js:60]: @xO-Tx Would be good to have a list of what happened before:
- If I type a URL which already exist, what happens? (does it show the redirect, does it tell you anything, etc)
- If I type a URL which is the same after the slugification (what you tried to do here)
- etc

My opinion without much thinking is that it's probably overkill to try to check if the URL already exist or not (I mean if it is changed automatically after save if it does exist, it seems ok). As for the slugification part `"/contact-us" to "contact us"` I am not sure it is as overkill as you think it is @rdeodoo, I can very well imagine people trying to replace the "-" inside the URL that was automatically added by spaces ^^ But I agree the implementation at least seems overkill (although probably correct even though incomplete as you said).

Anyways: I would really like a list of features that were in 15.0 and how it worked @xO-Tx :+1:

Comment (rdeodoo) [reply addons/website/static/src/components/fields/fields.js:60]: Well I didn't check if possible (is the reload occuring after a RPC based on that rpc result?) but maybe the python could simply slug and check if the slug is the same as the httprequest.request.path -> The slug ends up the same as the incoming url, don't reload, or even the JS could do that (if the reload url given by the server is the same as the current page url).

**Comment by bouvyd** [comment]
@qsm-odoo i suspect it would feel less weird if the focus was on the first field automatically:
- the underline of the field would be shown
- the blinking carret would make it clear that something is being edited

that's the behaviour in normal wizards, but not in yours
image
 
i would suggest to set the autofocus on the first field of your form modal and see if it makes it clearer

**Comment by xO-Tx** [comment]
> @xO-Tx Would be good to have a list of what happened before:
> - If I type a URL which already exist, what happens? (does it show the redirect, does it tell you anything, etc)
> - If I type a URL which is the same after the slugification (what you tried to do here)
>...

**A-** @qsm-odoo this is what we have in **15.0/16.0**:
There is no check in backend on URL input, just a JS check as on **15.0**:
```javascript
_onUrlChanged: function () {
    var url = this.$('input#page_url').val();
    this.$('.ask_for_redirect').toggleClass('d-none', url === this.page.url);
},
```
Let's suppose the URL of a website's home page is `'home'` and we have another page specific to the same website with URL = `'otherpage'`:

**1-** Type a URL that doesn't already exist on the same website but different than `'home'` (e.g. `'home/page'`) > redirect available > SAVE > URL changed to`'home/page'` (normal scenario)

**2-** Type something like `'home/'` or `'home '` > redirect available > SAVE > shouldn't change anything but in this case it will lead to a new URL = `'home-1'` **[B]**

**3-** Type invalid URL (e.g. `'home/new page/'`) > redirect enabled > SAVE > sets a unique URL slugified = `'home/new-page'`

**4-** Type a URL which already exist on the same website (`'otherpage'`) > redirect available > SAVE > unique url = `'otherpage-1'`

It's especially the **[A-2]** use case that I'm targeting here: instead of comparing the typed URL against current one, we can make things smarter and slugify first to prevent useless redirect... but now I found a strange behaviour on test : the submit of URL as in **[A-2]** will lead to a different value after "unique path check" (see **[B]**)

> The slugify is always done in python and always come with the unique path check, slugifying in js will always be incomplete and there is no way to compute the final url, as it needs to reach all other pages to ensure it is not already taken (and if so act accordingly, something like adding a suffix, I don't have code in front of me right now)

> Well I didn't check if possible (is the reload occuring after a RPC based on that rpc result?) but maybe the python could simply slug and check if the slug is the same as the httprequest.request.path -> The slug ends up the same as the incoming url, don't reload, or even the JS could do that (if the reload url given by the server is the same as the current page url).

**B-** @rdeodoo Yes I agree! but the goal here not to compute the final URL it's just to prevent in JS a basic scenario where the URL will be the same (after python slugify) but I found another issue while testing: I think the python code doesn't exclude the updated page while checking unique path which will lead to the result in **[A-2]**... I did a quick check now on the code (maybe you can confirm this :+1: ):

```python
def get_unique_path(self, page_url):
...
    while self.env['website.page'].with_context(active_test=False).sudo().search([('url', '=', page_temp)] + domain_static):
        inc += 1
        page_temp = page_url + (inc and "-%s" % inc or "")
...
```
This will consider the URL as duplicate if I submit the same URL after slugify... but in `'website.page' > write()` there is code to prevent it:
```python
def write(self, vals):
...
    if page.url != url: # HERE
        url = self.get_valid_page_url(url, website_id)
        page.menu_ids.write({'url': url})
...
```
Which I think won't work if `page.url == 'home'` and `url =='home/' `(since it compares them before slugify)... I think it can be fixed with something like this (To test...):
```diff
+  url = '/' + slugify(url, max_length=1024, path=True)
    if page.url != url:
        url = self.get_valid_page_url(url, website_id)
        page.menu_ids.write({'url': url})
...
```
So to sum up: The JS slugify in this PR only "Adds more occuracy" to prevent useless redirect as in **[A-2]**, but I think this works correctly only if we don't compare the page URL to itself in backend. @rdeodoo what do you think ?

> i would suggest to set the autofocus on the first field of your form modal and see if it makes it clearer


C- @qsm-odoo I think I can simply set the autofocus using `useAutofocus()` hook on JS field (even if personally I'm not convinced this is a generic solution since we still have strange "empty" selections and inputs on form view, **e.g. from a normal form view**:

field-empty

But Yes if there is a SPEC behind it I'm OK too :+1:

cc @bouvyd

**Comment by qsm-odoo** [comment]
@xO-Tx Good explanation, if what you found out on the python side is true I guess it could be worth a fix over there. For the JS side, maybe it's a bit overkill (mainly because of the need to make an RPC etc and the bunch of lines needed to achieve it) but I am not strongly against like @rdeodoo seemed to be, let's let him decide, he knows this flow and code better than me.

**Comment by qsm-odoo** [comment]
@bouvyd (cc @stefanorigano I guess you took part in that redesign) Indeed, I guess autofocus of the first field might help. But as @xO-Tx said above, the second tab of this dialog and some form view feels weird mainly because they are mostly empty fields by default :grimacing: 
image
(the floating caret here feels weird)
or @xO-Tx 's screenshot:
image
Fun fact, in community, the non required fields are worse than enterprise IMO (full border on hover) but the required fields are better:
image
 better than:
image

Maybe a very light gray background color on empty fields is the solution to all those problems here ?

**Comment by stefanorigano** [comment]
> cc @stefanorigano I guess you took part in that redesign)

Not really but I would be glad to help if I can :)

**TL;DR:** 
- Autofocus first field 👍 
- Add placeholders wherever possible, even a "_Not set_" for dropdowns would help
- Reduce fields' widths:  "visibility" shouldn't need ~100 characters 😉
- Reduce the modal's width: since we have always 1 column here, reduce the width will help the user to focus on a more compact layout.

**Some notes**:
> Maybe a very light gray background color on empty fields is the solution to all those problems here ?

 I'm against this kind of solutions for consistency reasons. We may find the perfect workaround for this modal, but instructing the user to be familiar with the UI and then propose him something different may confuse him even more ("_why grey bgs? are these fields disabled or what?_")
 
 **Design rant**:
Call me "old", but I have always find enterprise fields misleading and unpractical... for me a field is a field and a line is a line ^^
For v8 they told me that I would eventually get used to that,  but still... I find edition in ListView and community much simpler.
Who implemented the "always editable" behaviour made a miracle since it works flawlessly without worsening the UX... but it can't solve the original issue that (IMO) is enterprise fields design.

**Comment by rdeodoo** [comment]
@xO-Tx Not sure why didn't you continue the discussion on.. the discussion.. [here](https://github.com/odoo/odoo/pull/101022#discussion_r984068531), but instead replied in the middle of the thread were the discussion is already going elsewhere, but let's keep going here I guess :shrug: 

> There is no check in backend on URL input, just a JS check:

I am not sure of what we are talking about here TBH. There is python validation of the url input, which is actually checking for the slug and unique path. Maybe we are talking about something else.

About the commit message and what was already discussed, what kind of redirect are we talking about? Browser/js redirect ? Python redirect, or `website.rewrite` redirect?

> It's especially the [A-2] use case that I'm targeting here

So that's about the creation of the `website.rewrite`, right? I tested and indeed there is an useless redirect record created.
That's because the python code was not strong enough and not assuming that this case could exists: if the slugified url of the written url is the same as the already set url, it shouldn't create the redirect record.

So in 15.0, here:
```py
        # Create redirect if needed
        if data['create_redirect']:
            self.env['website.rewrite'].create({
                'name': data['name'],
                'redirect_type': data['redirect_type'],
                'url_from': original_url,
                'url_to': url,
                'website_id': website.id,
            })
``` 
in the `save_page_info()`, it should simply add a check `and original_url != url` with a comment `both url could be the same due to the slug` (better explained than that ofc)
This could be a 14.0 fix.

If that's what we are talking about, it shouldn't be a JS fix but just something better handled in the python/end point side, otherwise the issue and misconfiguration remains possible through other flows

Also, in master, we should introduce an sql constraint about this misconfiguration.

------------------------------------_

Now, reading your **point B**, after having checked and written :point_up: , I guess you were not talking about this issue actually (which should still be cared of I guess).

About that issue, I don't think adding `url = '/' + slugify(url, max_length=1024, path=True)` to check if the slugifies will result in the same url as the current one is ideal, most of the time we will do that twice for no reason (a first time to check that "error" you try to fix, and a second time to get the actual url if we are not in the case you mention).

At best maybe `get_valid_page_url` could take a new param of the pages but that's going too far TBH, unless someone disagrees.

We don't really care if someone rename it's page url from `/contactus` to `/contactus ` (or `/contactus/`) and it becomes `/contactus-1` instead of remaining `/contactus`, that wouldn't actually make more sense, would it? Still not what the user asked.

I think that's a non real use case and we shouldn't introduce code to handle that. It's already a bit complicated as is.

But now that I read the code, I am pretty sure I asked `get_valid_page_url()` to not be added as util but kept duplicate 2 lines of code in a previous PR.
That's probably the way to go here ;-)

```py
def write(self, vals):
    # ...
    if page.url != url:
        url = self.get_valid_page_url(url, website_id)
        page.menu_ids.write({'url': url})
    # ...

def clone_page(self, page_id, page_name=None, clone_menu=True):
    # ...
    if page_name:
        copy_param['url'] = self.get_valid_page_url(page_name)

def get_valid_page_url(self, page_url, website_id=False):
    url = '/' + slugify(page_url, max_length=1024, path=True)
    return self.env['website'].with_context(website_id=website_id).get_unique_path(url)
```

=>

```py
def write(self, vals):
    # ...
    url = url = '/' + slugify(url, max_length=1024, path=True)
    if page.url != url:
        url =  self.env['website'].with_context(website_id=website_id).get_unique_path(url)
        page.menu_ids.write({'url': url})
    # ...

def clone_page(self, page_id, page_name=None, clone_menu=True):
    # ...
    if page_name:
        url = '/' + slugify(page_name, max_length=1024, path=True)
        copy_param['url'] = self.env['website'].with_context(website_id=website_id).get_unique_path(url)
```

File: addons/website/static/src/components/fields/fields.js (hunk 1 above)
Comment (qsm-odoo) [reply addons/website/static/src/components/fields/fields.js:60]: @xO-Tx Good explanation, if what you found out on the python side is true I guess it could be worth a fix over there. For the JS side, maybe it's a bit overkill (mainly because of the need to make an RPC etc and the bunch of lines needed to achieve it) but I am not strongly against like @rdeodoo seemed to be, let's let him decide, he knows this flow and code better than me.

Comment (qsm-odoo) [reply addons/website/static/src/components/fields/fields.js:60]: @rdeodoo 's MESSAGE (MOVED BACK HERE):

> There is no check in backend on URL input, just a JS check:

I am not sure of what we are talking about here TBH. There is python validation of the url input, which is actually checking for the slug and unique path. Maybe we are talking about something else.

About the commit message and what was already discussed, what kind of redirect are we talking about? Browser/js redirect ? Python redirect, or `website.rewrite` redirect?

> It's especially the [A-2] use case that I'm targeting here

So that's about the creation of the `website.rewrite`, right? I tested and indeed there is an useless redirect record created.
That's because the python code was not strong enough and not assuming that this case could exists: if the slugified url of the written url is the same as the already set url, it shouldn't create the redirect record.

So in 15.0, here:
```py
        # Create redirect if needed
        if data['create_redirect']:
            self.env['website.rewrite'].create({
                'name': data['name'],
                'redirect_type': data['redirect_type'],
                'url_from': original_url,
                'url_to': url,
                'website_id': website.id,
            })
``` 
in the `save_page_info()`, it should simply add a check `and original_url != url` with a comment `both url could be the same due to the slug` (better explained than that ofc)
This could be a 14.0 fix.

If that's what we are talking about, it shouldn't be a JS fix but just something better handled in the python/end point side, otherwise the issue and misconfiguration remains possible through other flows

Also, in master, we should introduce an sql constraint about this misconfiguration.

------------------------------------_

Now, reading your **point B**, after having checked and written :point_up: , I guess you were not talking about this issue actually (which should still be cared of I guess).

About that issue, I don't think adding `url = '/' + slugify(url, max_length=1024, path=True)` to check if the slugifies will result in the same url as the current one is ideal, most of the time we will do that twice for no reason (a first time to check that "error" you try to fix, and a second time to get the actual url if we are not in the case you mention).

At best maybe `get_valid_page_url` could take a new param of the pages but that's going too far TBH, unless someone disagrees.

We don't really care if someone rename it's page url from `/contactus` to `/contactus ` (or `/contactus/`) and it becomes `/contactus-1` instead of remaining `/contactus`, that wouldn't actually make more sense, would it? Still not what the user asked.

I think that's a non real use case and we shouldn't introduce code to handle that. It's already a bit complicated as is.

But now that I read the code, I am pretty sure I asked `get_valid_page_url()` to not be added as util but kept duplicate 2 lines of code in a previous PR.
That's probably the way to go here ;-)

```py
def write(self, vals):
    # ...
    if page.url != url:
        url = self.get_valid_page_url(url, website_id)
        page.menu_ids.write({'url': url})
    # ...

def clone_page(self, page_id, page_name=None, clone_menu=True):
    # ...
    if page_name:
        copy_param['url'] = self.get_valid_page_url(page_name)

def get_valid_page_url(self, page_url, website_id=False):
    url = '/' + slugify(page_url, max_length=1024, path=True)
    return self.env['website'].with_context(website_id=website_id).get_unique_path(url)
```

=>

```py
def write(self, vals):
    # ...
    url = url = '/' + slugify(url, max_length=1024, path=True)
    if page.url != url:
        url =  self.env['website'].with_context(website_id=website_id).get_unique_path(url)
        page.menu_ids.write({'url': url})
    # ...

def clone_page(self, page_id, page_name=None, clone_menu=True):
    # ...
    if page_name:
        url = '/' + slugify(page_name, max_length=1024, path=True)
        copy_param['url'] = self.env['website'].with_context(website_id=website_id).get_unique_path(url)
```

Comment (qsm-odoo) [reply addons/website/static/src/components/fields/fields.js:60]: (no I am not talking to myself)

Comment (rdeodoo) [reply addons/website/static/src/components/fields/fields.js:60]: > There is no check in backend on URL input, just a JS check:

I am not sure of what we are talking about here TBH. There is python validation of the url input, which is actually checking for the slug and unique path. Maybe we are talking about something else.

About the commit message and what was already discussed, what kind of redirect are we talking about? Browser/js redirect ? Python redirect, or `website.rewrite` redirect?

> It's especially the [A-2] use case that I'm targeting here

So that's about the creation of the `website.rewrite`, right? I tested and indeed there is an useless redirect record created.
That's because the python code was not strong enough and not assuming that this case could exists: if the slugified url of the written url is the same as the already set url, it shouldn't create the redirect record.

So in 15.0, here:
```py
        # Create redirect if needed
        if data['create_redirect']:
            self.env['website.rewrite'].create({
                'name': data['name'],
                'redirect_type': data['redirect_type'],
                'url_from': original_url,
                'url_to': url,
                'website_id': website.id,
            })
``` 
in the `save_page_info()`, it should simply add a check `and original_url != url` with a comment `both url could be the same due to the slug` (better explained than that ofc)
This could be a 14.0 fix.

If that's what we are talking about, it shouldn't be a JS fix but just something better handled in the python/end point side, otherwise the issue and misconfiguration remains possible through other flows

Also, in master, we should introduce an sql constraint about this misconfiguration.

------------------------------------_

Now, reading your **point B**, after having checked and written :point_up: , I guess you were not talking about this issue actually (which should still be cared of I guess).

About that issue, I don't think adding `url = '/' + slugify(url, max_length=1024, path=True)` to check if the slugifies will result in the same url as the current one is ideal, most of the time we will do that twice for no reason (a first time to check that "error" you try to fix, and a second time to get the actual url if we are not in the case you mention).

At best maybe `get_valid_page_url` could take a new param of the pages but that's going too far TBH, unless someone disagrees.

We don't really care if someone rename it's page url from `/contactus` to `/contactus ` (or `/contactus/`) and it becomes `/contactus-1` instead of remaining `/contactus`, that wouldn't actually make more sense, would it? Still not what the user asked.

I think that's a non real use case and we shouldn't introduce code to handle that. It's already a bit complicated as is.

But now that I read the code, I am pretty sure I asked `get_valid_page_url()` to not be added as util but kept duplicate 2 lines of code in a previous PR.
That's probably the way to go here ;-)

```py
def write(self, vals):
    # ...
    if page.url != url:
        url = self.get_valid_page_url(url, website_id)
        page.menu_ids.write({'url': url})
    # ...

def clone_page(self, page_id, page_name=None, clone_menu=True):
    # ...
    if page_name:
        copy_param['url'] = self.get_valid_page_url(page_name)

def get_valid_page_url(self, page_url, website_id=False):
    url = '/' + slugify(page_url, max_length=1024, path=True)
    return self.env['website'].with_context(website_id=website_id).get_unique_path(url)
```

=>

```py
def write(self, vals):
    # ...
    url = url = '/' + slugify(url, max_length=1024, path=True)
    if page.url != url:
        url =  self.env['website'].with_context(website_id=website_id).get_unique_path(url)
        page.menu_ids.write({'url': url})
    # ...

def clone_page(self, page_id, page_name=None, clone_menu=True):
    # ...
    if page_name:
        url = '/' + slugify(page_name, max_length=1024, path=True)
        copy_param['url'] = self.env['website'].with_context(website_id=website_id).get_unique_path(url)
```

Comment (rdeodoo) [reply addons/website/static/src/components/fields/fields.js:60]: @qsm-odoo I deleted the last one and repasted it, will make stuff a (little) bit clearer :joy: 

@xO-Tx Try to replay on discussion on.. discussion, here you started to replay to everyone in a message in the tread, instead of continuing here, it was impossible to understand anything anymore. Thanks @qsm-odoo for moving stuff here.

Comment (xO-Tx) [reply addons/website/static/src/components/fields/fields.js:60]: > in the `save_page_info()`, it should simply add a check `and original_url != url` with a comment `both url could be the same due to the slug` (better explained than that ofc)
> This could be a 14.0 fix.
> 
> If that's what we are talking about, it shouldn't be a JS fix but just something better handled in the python/end point side, otherwise the issue and misconfiguration remains possible through other flows

@rdeodoo actually that's the useless `website.rewrite` I was talking about and I added this "JS slugify version" mainly because I saw it was confusing to allow it in first place by making the **"page properties redirect option"** visible while it shouldn't be.

> We don't really care if someone rename it's page url from `/contactus` to `/contactus ` (or `/contactus/`) and it becomes `/contactus-1` instead of remaining `/contactus`, that wouldn't actually make more sense, would it? Still not what the user asked.

Mmm... I thought it was a consistency issue... because this changes everything now:

**C1-** We want `/contactus/` or ... to be `/contactus`:
- We should prevent backend from setting `/contactus-1`
- We should prevent backend from creating a `website.rewrite` to the same URL.
- We should disable the "redirect" option in page properties.

**C2-** It's OK to have `/contactus/` or ... => `/contactus-1`
- It's a different URL, so we can create a `website.rewrite` for it (no problem here).
- Nothing to do on JS.

My goal here was to have the configuration in **C1** but if **C2** is OK for you I'm OK too :+1: 

> But now that I read the code, I am pretty sure I asked get_valid_page_url() to not be added as util but kept duplicate 2 lines of code in a previous PR.

I remember that I tried many times to remove it, I think the last time was a part of the [wizard PR](https://github.com/odoo/odoo/pull/95949/commits/275c04bbb25e9d5900aed651304ea5986aa59507#diff-77bd6b19c39e211959885024bcad914655ff84cfc10c16633687d014e50aa69aL276), but I don't think I removed in other PR after we decided to close it :grimacing:

> ```py
> def write(self, vals):
>     # ...
>     url = url = '/' + slugify(url, max_length=1024, path=True)
>     if page.url != url:
>         url =  self.env['website'].with_context(website_id=website_id).get_unique_path(url)
>         page.menu_ids.write({'url': url})
>     # ...
> 
> def clone_page(self, page_id, page_name=None, clone_menu=True):
>     # ...
>     if page_name:
>         url = '/' + slugify(page_name, max_length=1024, path=True)
>         copy_param['url'] = self.env['website'].with_context(website_id=website_id).get_unique_path(url)
> ```

Indeed, it's better this way :+1:

Comment (xO-Tx) [reply addons/website/static/src/components/fields/fields.js]: @qsm-odoo I added a small tweak on the code:
- Remove the **"JS slugify code"** and keep only the **"/ in non edited part of the field"**
- Remove the `get_valid_page_url()` (as @rdeodoo suggested) & set focus on the first field in page properties :+1:

File: addons/website/static/src/components/fields/fields.js [hunk 2]
```diff
@@ -16,15 +16,21 @@ const {Component, useState, onWillStart} = owl;
...
+            url: this.pageUrl,
             redirect_type: '301',
         });
-        useInputField({getValue: () => this.props.value.url || this.props.value});
 
-        this.serverUrl = window.location.origin;
-        this.pageUrl = this.props.value;
+        onWillStart(async () => {
+            this.pageMimetypes = await this.orm.call('website', 'guess_mimetype', []);
+        });
```
Comment (rdeodoo) [inline addons/website/static/src/components/fields/fields.js:32]: What is this `pageMimetypes` used for?

**Review by rdeodoo — CHANGES_REQUESTED** [review]
There is actually a problem with the current implementation, you can't set back a homepage (You would simply empty the field, which is actually required, it won't let you).
One could type `/` in the input but that wouldn't be ideal as opposed to other flows where you don't have to type the leading slash (that you remove anyway) + it would show a double // (one in the domain, one in the path).

From there:
- Either make that input not required (required=0 on the form view with a comment explaining that despite required, it's not needed as the / will be added by the write method)
- Either change back that implementation to keep the / outside (and maybe do the opposite of the current regex -> Add the leading slash if not typed by the user).

I let you decide which one you prefer and think makes more sense UX wise for an end user.

File: addons/website/static/src/components/fields/fields.js (hunk 2 above)
Comment (xO-Tx) [reply addons/website/static/src/components/fields/fields.js]: Thanks for spotting this! this is a dead code now since I removed JS slugify... removed :+1:

**Comment by xO-Tx** [comment]
> * Either make that input not required (required=0 on the form view with a comment explaining that despite required, it's not needed as the / will be added by the write method)
> * Either change back that implementation to keep the / outside (and maybe do the opposite of the current regex -> Add the leading slash if not typed by the user).
> 
> I let you decide which one you prefer and think makes more sense UX wise for an end user.

I don't think `required="0"` will work, we'll get an error since the `write()` can't handle value = `False` ([I added a FIX for this on 13.0](https://github.com/odoo/odoo/pull/98501)... sorry for the delay, I answered now your comment on the PR :+1:).

So keeping the URL "required" (and even adding `required=True` on `website.page` in python model) makes sense, but in terms of UX, the best option IMHO is to just fix `write()` to deal with `False` as in the FIX and remove "required" form the URL as you suggested... If this configuration is OK for you, you can fwport #98501 so I can remove the "required" from the URL here :+1:

File: addons/website/static/src/components/fields/fields.js [hunk 3]
```diff
@@ -36,6 +38,10 @@ class PageUrlField extends Component {
         this.updateValues();
     }
 
+    get fieldURL() {
+        return (this.props.value.url || this.props.value).replace(/^\//g, '');
```
Comment (rdeodoo) [inline addons/website/static/src/components/fields/fields.js:42]: I don't get this fallback on this.props.value it is setting a whole object as url ?
image

- Go to any page
- Empty the URL
- Toggle the redirect dropdown

**Review by rdeodoo — CHANGES_REQUESTED** [review]
> the best option IMHO is to just fix write() to deal with False as in the FIX and remove "required" form the URL as you suggested... If this configuration is OK for you, you can fwport https://github.com/odoo/odoo/pull/98501 so I can remove the "required" from the URL here

Ok let's do that, I agree. It's too late to add `required=True` on the `url` field definition anyway.
Note that I didn't merge the 13.0 PR, too risky, see my message on it.

For this PR, you can simply add `or ''` as fallback on this line: https://github.com/odoo/odoo/blame/c25609b1d9ac172df40d77884807a3c1cdfbb584/addons/website/models/website_page.py#L161 and indeed remove the required, but something else need to be done, see my review.

If you can adapt that tomorrow morning and I'll merge it :+1:

File: addons/website/static/src/components/fields/fields.js (hunk 3 above)
Comment (xO-Tx) [reply addons/website/static/src/components/fields/fields.js]: From what I see in the screenshot, you were testing on a '16.0' runbot instance: https://20117364-16-0-all.runbot29.odoo.com/ (not this PR's code :thinking:), but indeed, this flow should be fixed (it leads to a different traceback with the current code too...) fixed now, thanks for spotting this :+1:

Comment (rdeodoo) [reply addons/website/static/src/components/fields/fields.js]: I was not, I just removed the required=True locally on top of your PR, I couldn't have tested that line on the regular runbot as it doesn't exists :p

**Review by rdeodoo — APPROVED** [review]
@robodoo r+
Thanks for the overnight adaptation :+1:
